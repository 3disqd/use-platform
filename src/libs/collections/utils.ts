import { Fragment, ReactElement, ReactNode, isValidElement } from 'react'

import { CollectionComponent, Renderer } from './types'

export function isReactFragment(node: ReactNode): node is ReactElement<{ children?: ReactNode }> {
  return isValidElement(node) && node.type === Fragment
}

export function isCollectionElement<T = unknown, P = {}>(
  node: ReactNode,
): node is ReactElement<P, CollectionComponent<T, P>> {
  return (
    isValidElement(node) &&
    typeof node.type === 'function' &&
    typeof (node.type as CollectionComponent).getCollectionNode === 'function'
  )
}

// TODO: Подумать про кейс с ReactFragment и как избавиться от этой функции
export function isRenderer<T>(value: unknown): value is Renderer<T> {
  return typeof value === 'function'
}
