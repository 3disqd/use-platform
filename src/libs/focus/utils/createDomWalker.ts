export function createDomWalker(root: Node, predicate: (node: Node) => boolean) {
  const acceptNode: NodeFilter['acceptNode'] = (node) => {
    if (predicate(node)) {
      return NodeFilter.FILTER_ACCEPT
    }

    return NodeFilter.FILTER_SKIP
  }

  // IE11 require use function instead object.
  const safeFilter = acceptNode as unknown as NodeFilter
  safeFilter.acceptNode = acceptNode

  const walker = document.createTreeWalker(
    root,
    NodeFilter.SHOW_ELEMENT,
    safeFilter,
    // IE11 require last argument
    false,
  )

  return walker
}
