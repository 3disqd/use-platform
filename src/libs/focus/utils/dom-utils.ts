let protoMatches: Element['matches'] | null = null
if (typeof Element !== 'undefined') {
  protoMatches =
    Element.prototype.matches ||
    Element.prototype.webkitMatchesSelector ||
    // @ts-expect-error
    Element.prototype.matchesSelector ||
    // @ts-expect-error
    Element.prototype.mozMatchesSelector ||
    // @ts-expect-error
    Element.prototype.msMatchesSelector
}

export function matches(node: Element, selectors: string) {
  return protoMatches ? protoMatches.call(node, selectors) : false
}

export function isInput(node: HTMLElement): node is HTMLInputElement {
  return node.tagName === 'INPUT'
}

export function isIframe(node: HTMLElement): node is HTMLFrameElement {
  return node.tagName === 'IFRAME'
}

export function isHiddenInput(node: HTMLElement): node is HTMLInputElement {
  return isInput(node) && node.type === 'hidden'
}

export function isRadioInput(node: HTMLElement): node is HTMLInputElement {
  return isInput(node) && node.type === 'radio'
}

export function isHidden(node: Element) {
  if (window.getComputedStyle(node).visibility === 'hidden') {
    return true
  }

  let parent: Element | null = node
  while (parent) {
    if (window.getComputedStyle(parent).display === 'none') {
      return true
    }

    parent = parent.parentElement
  }

  return false
}
