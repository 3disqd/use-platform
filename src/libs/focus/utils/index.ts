export * from './createDomWalker'
export * from './dom-utils'
