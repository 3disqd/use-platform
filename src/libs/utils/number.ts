export function clamp(value: number, min = -Infinity, max = Infinity) {
  return Math.min(Math.max(value, min), max)
}

export function isInRange(value: number, min: number, max: number) {
  return value >= min && value <= max
}
