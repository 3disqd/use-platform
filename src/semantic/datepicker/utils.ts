import { DateValue } from '../../shared/types'

/**
 * Uses time from second date for first (if first hasn't time).
 */
export function getDateWithTime(date: DateValue, time: DateValue): Date {
  const nextDate = new Date(date)

  const hasTime = Boolean(
    nextDate.getHours() ||
      nextDate.getMinutes() ||
      nextDate.getSeconds() ||
      nextDate.getMilliseconds(),
  )

  if (hasTime) {
    return new Date(date)
  }

  time = new Date(time)

  nextDate.setHours(time.getHours(), time.getMinutes(), time.getSeconds(), time.getMilliseconds())

  return nextDate
}
