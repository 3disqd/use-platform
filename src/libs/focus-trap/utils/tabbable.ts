import { ElementTreeWalker, isTabbable, adjustedTabIndex } from '../../focus'

interface TabbableNode {
  index: number
  tabIndex: number
  node: HTMLElement
}

export function getTabbables(root: HTMLElement, filter?: (node: HTMLElement) => boolean) {
  const walker = new ElementTreeWalker(root, (node) => {
    const tabbable = isTabbable(node)

    if (filter && tabbable) {
      return filter(node)
    }

    return tabbable
  })

  const nodes: TabbableNode[] = []
  const regular: HTMLElement[] = []

  let index = 0
  let node: HTMLElement | null = null
  while ((node = walker.next())) {
    const tabIndex = adjustedTabIndex(node)

    if (tabIndex === 0) {
      regular.push(node)
    } else {
      nodes.push({
        index: index++,
        tabIndex,
        node,
      })
    }
  }

  return nodes
    .sort((a, b) => (a.tabIndex === b.tabIndex ? a.index - b.index : a.tabIndex - b.tabIndex))
    .map((v) => v.node)
    .concat(regular)
}
