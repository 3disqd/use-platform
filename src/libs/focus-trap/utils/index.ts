import { focusElement } from '../../dom-utils'
import { getTabbables } from './tabbable'

const FOCUS_TRAP_GUARD_ATTR = 'focusTrapGuard'

export function getActiveElement() {
  return document.activeElement as HTMLElement | null
}

export function isElementInScope(scope: Element, element: Element) {
  return scope.contains(element)
}

export function isFocusFree() {
  return getActiveElement() === document.body
}

export function createGuard(tabIndex: number) {
  const element = document.createElement('span')
  element.tabIndex = tabIndex
  Object.assign(element.style, {
    width: '1px',
    height: 0,
    padding: 0,
    overflow: 'hidden',
    position: 'fixed',
    top: '1px',
    left: '1px',
  })
  element.setAttribute('aria-hidden', 'true')
  element.dataset[FOCUS_TRAP_GUARD_ATTR] = 'true'

  return element
}

export function isGuard(node: HTMLElement) {
  return Boolean(node.dataset[FOCUS_TRAP_GUARD_ATTR])
}

export function removeElementFromDocument(element: Element) {
  if (!element.parentNode) {
    return
  }

  element.parentNode.removeChild(element)
}

export function focusFirstIn(scope: HTMLElement) {
  const [first] = getTabbables(scope, (node) => !isGuard(node))

  focusElement(first)
}

export function getNextTabbableIndex(current: number, count: number, step: number) {
  if (current === -1) {
    return 0
  }

  return count > 0 ? (count + current + step) % count : -1
}
