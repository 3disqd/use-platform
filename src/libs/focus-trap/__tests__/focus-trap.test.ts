import { getByTestId } from '@testing-library/react'

import { getTabbables } from '../utils/tabbable'
import { createGuard, isFocusFree, isGuard, focusFirstIn, getNextTabbableIndex } from '../utils'
import { basicFixture } from './fixtures'

describe('focus-trap', () => {
  describe('isFocusFree', () => {
    test('should return true if body is focused`', () => {
      const focusable = document.createElement('button')
      document.body.appendChild(focusable)

      focusable.focus()
      expect(isFocusFree()).toBe(false)

      focusable.blur()
      expect(isFocusFree()).toBe(true)

      document.body.removeChild(focusable)
    })
  })

  describe('isGuard', () => {
    test('should return true if element is guard', () => {
      const focusable = document.createElement('button')
      expect(isGuard(focusable)).toBe(false)

      const guard = createGuard(0)
      expect(isGuard(guard)).toBe(true)
    })
  })

  describe('focusFirstIn', () => {
    test('should set focus in first tabbable element', () => {
      const guard = createGuard(1)
      const tabbable = document.createElement('button')
      document.body.appendChild(guard)
      document.body.appendChild(tabbable)

      focusFirstIn(document.body)
      expect(tabbable).toHaveFocus()

      document.body.removeChild(guard)
      document.body.removeChild(tabbable)
    })
  })

  describe('getNextTabbableIndex', () => {
    test('should return first index if current is "-1"', () => {
      expect(getNextTabbableIndex(-1, 5, 0)).toBe(0)
    })

    test('should return fist index if current is last index (step=1)', () => {
      expect(getNextTabbableIndex(4, 5, 1)).toBe(0)
    })

    test('should return last index if current is first index (step=-1)', () => {
      expect(getNextTabbableIndex(0, 5, -1)).toBe(4)
    })

    test('should return "-1" if count=0', () => {
      expect(getNextTabbableIndex(20, 0, 1)).toBe(-1)
    })

    test('should return next relative index', () => {
      expect(getNextTabbableIndex(5, 10, 1)).toBe(6)
      expect(getNextTabbableIndex(5, 10, -1)).toBe(4)
    })
  })

  describe('getTabbables', () => {
    test('should return tabbable-elements inside container', () => {
      const wrapper = document.createElement('div')
      wrapper.innerHTML = basicFixture
      document.body.appendChild(wrapper)

      // https://github.com/jsdom/jsdom/issues/1670
      const divContentEditableTrue = getByTestId(wrapper, 'div-contenteditable-true')
      const divContentEditableFalse = getByTestId(wrapper, 'div-contenteditable-false')
      const divContentEditableNesting = getByTestId(wrapper, 'div-contenteditable-nesting')
      divContentEditableTrue.contentEditable = 'true'
      divContentEditableFalse.contentEditable = 'false'
      divContentEditableNesting.contentEditable = 'true'

      const tabbables = getTabbables(wrapper)

      expect(tabbables.map((node) => node.dataset.testid)).toMatchSnapshot()

      document.body.removeChild(wrapper)
    })

    test('should return filtered tabbable-elements', () => {
      const wrapper = document.createElement('div')
      wrapper.innerHTML = basicFixture
      document.body.appendChild(wrapper)

      const tabbables = getTabbables(wrapper, (node) => node.tagName === 'BUTTON')

      expect(tabbables.map((node) => node.dataset.testid)).toMatchSnapshot()

      document.body.removeChild(wrapper)
    })
  })
})
