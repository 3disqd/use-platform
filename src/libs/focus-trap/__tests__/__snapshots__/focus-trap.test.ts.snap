// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`focus-trap getTabbables should return filtered tabbable-elements 1`] = `
Array [
  "button",
  "button-visible-parent-hidden",
]
`;

exports[`focus-trap getTabbables should return tabbable-elements inside container 1`] = `
Array [
  "link-tabindex",
  "input",
  "input-readonly",
  "textarea",
  "textarea-readonly",
  "select",
  "select-readonly",
  "link-anchor",
  "button",
  "audio-control",
  "video-control",
  "div-tabindex",
  "div-contenteditable-true",
  "div-contenteditable-nesting",
  "button-visible-parent-hidden",
]
`;
